File: 501-FindModeinBinarySearchTree/501-FindModeinBinarySearchTree.java
/*

501. Find Mode in Binary Search Tree
Easy

Given a binary search tree (BST) with duplicates, find all the mode(s) (the most frequently occurred element) in the given BST.

Assume a BST is defined as follows:

    The left subtree of a node contains only nodes with keys less than or equal to the node's key.
    The right subtree of a node contains only nodes with keys greater than or equal to the node's key.
    Both the left and right subtrees must also be binary search trees.

 

For example:
Given BST [1,null,2,2],

   1
    \
     2
    /
   2


return [2].

Note: If a tree has more than one mode, you can return them in any order.

Follow up: Could you do that without using any extra space? (Assume that the implicit stack space incurred due to recursion does not count).

*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] findMode(TreeNode root) {
        Map<Integer, Integer> lookupMap = new HashMap<>();
        dfs(root, lookupMap);
        List<Integer> resList = new ArrayList<>();
        final Map<Integer, Integer> sorted = lookupMap.entrySet()
                .stream()
                .sorted((Map.Entry.<Integer, Integer>comparingByValue().reversed()))                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
        int max = 0;
        for(Map.Entry<Integer, Integer> entry : sorted.entrySet()){
            if (max <= entry.getValue()){
                resList.add(entry.getKey());
                max = entry.getValue();
            }
        }
        return resList.stream().mapToInt(i -> i).toArray();
        
    }
    private void dfs(TreeNode node, Map<Integer, Integer> lookupMap){
        if (node == null)
            return;
        lookupMap.put(node.val, lookupMap.containsKey(node.val) ? lookupMap.get(node.val) + 1 : 1);
        dfs(node.left, lookupMap);
        dfs(node.right, lookupMap);
    }
}

File: 501-FindModeinBinarySearchTree/501-FindModeinBinarySearchTree.py
"""

501. Find Mode in Binary Search Tree

Given the root of a binary search tree (BST) with duplicates, return all the mode(s) (i.e., the most frequently occurred element) in it.

If the tree has more than one mode, return them in any order.

Assume a BST is defined as follows:

    The left subtree of a node contains only nodes with keys less than or equal to the node's key.
    The right subtree of a node contains only nodes with keys greater than or equal to the node's key.
    Both the left and right subtrees must also be binary search trees.

Example 1:

Input: root = [1,null,2,2]
Output: [2]

Example 2:

Input: root = [0]
Output: [0]

Constraints:

    The number of nodes in the tree is in the range [1, 104].
    -105 <= Node.val <= 105

 
Follow up: Could you do that without using any extra space? (Assume that the implicit stack space incurred due to recursion does not count).

"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        def dfs(node):
            if not node: return
            dfs(node.left)
            res.append(node.val)
            dfs(node.right)
        res = []
        dfs(root)
        mapp = Counter(res)
        heap = []
        ans = []
        for k, v in mapp.items():
            heappush(heap, (-v, k))
        max_val = 0
        while heap:
            v, k = heappop(heap)
            v = -v
            if max_val > v: return ans
            max_val = v
            ans.append(k)
        return ans
